File: UCSample/UCSample/SoundService.m
#import "SoundService.h"
#import "AppDelegate.h"

#import <AVFoundation/AVFoundation.h>

// private implementation
//
@interface SoundService (Private)
+ (AVAudioPlayer *)initPlayerWithPath:(NSString *)path;
@end

@implementation SoundService (Private)

+ (AVAudioPlayer *)initPlayerWithPath:(NSString *)path {
  NSURL *url = [NSURL
      fileURLWithPath:[NSString
                          stringWithFormat:@"%@/%@",
                                           [[NSBundle mainBundle] resourcePath],
                                           path]];

  NSError *error;
  AVAudioPlayer *player =
      [[AVAudioPlayer alloc] initWithContentsOfURL:url error:&error];
  if (player == nil) {
  }

  return player;
}

@end

@interface SoundService () {
@private
  AVAudioPlayer *playerRingBackTone;
  AVAudioPlayer *playerRingTone;

  BOOL speakerOn;
}
@end
//
// default implementation
//
@implementation SoundService

- (SoundService *)init {
  if ((self = [super init])) {
  }
  return self;
}

- (void)dealloc {

#define RELEASE_PLAYER(player)                                                 \
  if (player) {                                                                \
    if (player.playing) {                                                      \
      [player stop];                                                           \
    }                                                                          \
  }

  RELEASE_PLAYER(playerRingBackTone);
  RELEASE_PLAYER(playerRingTone);

#undef RELEASE_PLAYER
}

//
// SoundService
//
- (BOOL)speakerEnabled:(BOOL)enabled {
  AVAudioSession *session = [AVAudioSession sharedInstance];
  AVAudioSessionCategoryOptions options = session.categoryOptions;

  if (enabled) {
    options |= AVAudioSessionCategoryOptionDefaultToSpeaker;
  } else {
    options &= ~AVAudioSessionCategoryOptionDefaultToSpeaker;
  }

  NSError *error = nil;
  [session setCategory:AVAudioSessionCategoryPlayAndRecord
           withOptions:options
                 error:&error];
  if (error != nil) {
    return NO;
  }
  return YES;
}

- (BOOL)isSpeakerEnabled {
  return speakerOn;
}

- (BOOL)playRingTone {
  if (!playerRingTone) {
    playerRingTone = [SoundService initPlayerWithPath:@"ringtone.mp3"];
  }
  if (playerRingTone) {
    playerRingTone.numberOfLoops = -1;
    [self speakerEnabled:YES];
    [playerRingTone play];
    return YES;
  }
  return NO;
}

- (BOOL)stopRingTone {
  if (playerRingTone && playerRingTone.playing) {
    [playerRingTone stop];
  }
  return YES;
}

- (BOOL)playRingBackTone {
  if (!playerRingBackTone) {
    playerRingBackTone = [SoundService initPlayerWithPath:@"ringtone.mp3"];
  }
  if (playerRingBackTone) {
    playerRingBackTone.numberOfLoops = -1;
    [self speakerEnabled:NO];
    [playerRingBackTone play];
    return YES;
  }

  return NO;
}

- (BOOL)stopRingBackTone {
  if (playerRingBackTone && playerRingBackTone.playing) {
    [playerRingBackTone stop];
  }
  return YES;
}

@end

File: UCSample_swift/UCSample/SoundService.swift
import AVFoundation

// private implementation
//
class SoundService {
    var playerRingBackTone: AVAudioPlayer!
    var playerRingTone: AVAudioPlayer!
    var speakerOn: Bool!

    func initPlayerWithPath(_ path: String) -> AVAudioPlayer {
        let url = URL(fileURLWithPath: Bundle.main.path(forResource: path, ofType: nil)!)

        var player: AVAudioPlayer!
        do {
            player = try AVAudioPlayer(contentsOf: url)
        } catch {}

        return player
    }

    func unInit() {
        if playerRingBackTone != nil {
            if playerRingBackTone.isPlaying {
                playerRingBackTone.stop()
            }
        }

        if playerRingTone != nil {
            if playerRingTone.isPlaying {
                playerRingTone.stop()
            }
        }
    }

    //
    // SoundService
    //
    func speakerEnabled(_ enabled: Bool) {
        let session = AVAudioSession.sharedInstance()
        var options = session.categoryOptions

        if enabled {
            options.insert(AVAudioSession.CategoryOptions.defaultToSpeaker)
        } else {
            options.remove(AVAudioSession.CategoryOptions.defaultToSpeaker)
        }
        do {
            try session.setCategory(AVAudioSession.Category.playAndRecord, options: options)
            NSLog("Playback OK")
        } catch {
            NSLog("ERROR: CANNOT speakerEnabled. Message from code: \"\(error)\"")
        }
        
        
    }

    func isSpeakerEnabled() -> Bool {
        speakerOn
    }

    func playRingTone() -> Bool {
        if playerRingTone == nil {
            playerRingTone = initPlayerWithPath("ringtone.mp3")
        }
        if playerRingTone != nil {
            playerRingTone.numberOfLoops = -1
            speakerEnabled(true)
            playerRingTone.play()
            return true
        }
        return false
    }

    func stopRingTone() -> Bool {
        if playerRingTone != nil, playerRingTone.isPlaying {
            playerRingTone.stop()
            speakerEnabled(true)
        }
        return true
    }

    func playRingBackTone() -> Bool {
        if playerRingBackTone == nil {
            playerRingBackTone = initPlayerWithPath("ringtone.mp3")
        }
        if playerRingBackTone != nil {
            playerRingBackTone.numberOfLoops = -1
            speakerEnabled(false)
            playerRingBackTone.play()
            return true
        }

        return false
    }

    func stopRingBackTone() -> Bool {
        if playerRingBackTone != nil, playerRingBackTone.isPlaying {
            playerRingBackTone.stop()
            speakerEnabled(true)
        }
        return true
    }
}
